daphne/src/messages.rs

```diff
@@ -809,6 +822,266 @@ impl Decode for HpkeCiphertext {
     }
 }
 
+// VdafType plus the variable-length portions associated with it, from draft-wang-ppm-dap-taskprov-00
+#[derive(Clone, Deserialize, Serialize, Debug, PartialEq, Eq)]
+pub enum VdafTypeVar {
+    Prio3Aes128Count,
+    Prio3Aes128Sum { bit_length: u8 },
+    Prio3Aes128Histogram { buckets: Vec<u64> },
+    Poplar1Aes128 { bit_length: u16 },
+    NotImplemented(u32),
+}
+
+impl Encode for VdafTypeVar {
+    fn encode(&self, bytes: &mut Vec<u8>) {
+        match &self {
+            VdafTypeVar::Prio3Aes128Count => VDAF_TYPE_PRIO3_AES128_COUNT.encode(bytes),
+            VdafTypeVar::Prio3Aes128Sum { bit_length } => {
+                VDAF_TYPE_PRIO3_AES128_SUM.encode(bytes);
+                bit_length.encode(bytes);
+            }
+            VdafTypeVar::Prio3Aes128Histogram { buckets } => {
+                VDAF_TYPE_PRIO3_AES128_HISTOGRAM.encode(bytes);
+                encode_u24_items(bytes, &(), buckets);
+            }
+            VdafTypeVar::Poplar1Aes128 { bit_length } => {
+                VDAF_TYPE_POPLAR1_AES128.encode(bytes);
+                bit_length.encode(bytes);
+            }
+            VdafTypeVar::NotImplemented(x) => {
+                x.encode(bytes);
+            }
+        }
+    }
+}
+
+impl Decode for VdafTypeVar {
+    fn decode(bytes: &mut Cursor<&[u8]>) -> Result<Self, CodecError> {
+        let x = u32::decode(bytes)?;
+        match x {
+            VDAF_TYPE_PRIO3_AES128_COUNT => Ok(Self::Prio3Aes128Count),
+            VDAF_TYPE_PRIO3_AES128_SUM => Ok(Self::Prio3Aes128Sum {
+                bit_length: u8::decode(bytes)?,
+            }),
+            VDAF_TYPE_PRIO3_AES128_HISTOGRAM => Ok(Self::Prio3Aes128Histogram {
+                buckets: decode_u24_items(&(), bytes)?,
+            }),
+            VDAF_TYPE_POPLAR1_AES128 => Ok(Self::Poplar1Aes128 {
+                bit_length: u16::decode(bytes)?,
+            }),
+            _ => Ok(Self::NotImplemented(x)),
```

Comment:
Consider the following situation: We're decoding a message like

```
struct {
   ...
   VdafConfig vdaf_config;
   Foo foo;
} Bar;
```

If we don't recognize the VdafType in VdafConfig, then we don't know where vdaf_config ends and foo begins. In my opinion, the only thing we can do here is fail.
```suggestion
            _ => Err(CodecError::UnexpectedValue),
```


I think we should err here instead. (And we should definitely fix this in the spec.)

Reply:
If we do fix the spec, e.g. by making the VdafConfig a length-prefixed thing, then we can come back and change this again.  But I agree that what you suggest is better until then.